### PR TITLE
Iterator

### PR DESCRIPTION
**Pull request by kilohsakul**
Classes providing an iterator that iterates through a vector of ordered vector synchrnounously. Can be used in determinisation, post, afa non-determinization.
Determinization and intersection rewritten using this iterator.

And I also boldly renamed TransSymbolsStates to Move.

### CODE REVIEW

**Comment by tfiedor** [comment]
Has someone fed me an acid, or what the fuck is going on? :D

File: 3rdparty/re2/re2/regexp.cc [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 namespace re2 {
 
-// Constructor.  Allocates vectors as appropriate for operator.
+// Constructor.  Allocates _vectors as appropriate for operator.
```
Comment (martinhruska) [inline 3rdparty/re2/re2/regexp.cc:26]: Is a real motivation to rename it from vectors to _vectors or is it juser unfortunate replace all by IDE.

File: AUTOMATAFORMAT.md [hunk 2]
```diff
@@ -45,12 +45,12 @@ We use the general form of AFA, by Pavol V., and simple NFA.
 * `true` means true and `false` means false.
 
 ## Alphabet type
-Next, we categorise automata by the **alphabet type**, i.e., the representation of symbols on transitions. We will consider propositional formulae (over bit vectors), explicit symbols, intervals of numbers or unicode or ascii, or user-defined alphabet. The type of alphabet is encoded in a name of section, e.g., `@NFA-explicit` denotes that nondeterministic finite automaton over explicit alphabet is given.
+Next, we categorise automata by the **alphabet type**, i.e., the representation of symbols on transitions. We will consider propositional formulae (over bit _vectors), explicit symbols, intervals of numbers or unicode or ascii, or user-defined alphabet. The type of alphabet is encoded in a name of section, e.g., `@NFA-explicit` denotes that nondeterministic finite automaton over explicit alphabet is given.
```
Comment (martinhruska) [inline AUTOMATAFORMAT.md:48]: Again, it looks like an unwanted replace all.

File: bindings/python/mata.pxd [hunk 3]
```diff
@@ -93,8 +93,8 @@ cdef extern from "mata/nfa.hh" namespace "Mata::Nfa":
     ctypedef umap[State, State] StateToStateMap
     ctypedef umap[Symbol, string] SymbolToStringMap
     ctypedef umap[string, string] StringDict
-    ctypedef COrdVector[CTransSymbolStates] TransitionList
-    ctypedef vector[TransitionList] TransitionRelation
+    ctypedef COrdVector[CTransSymbolStates] MoveList
+    ctypedef vector[MoveList] TransitionRelation
```
Comment (martinhruska) [inline bindings/python/mata.pxd:97]: Why is this renamed?

File: include/mata/inter-aut.hh [hunk 4]
```diff
@@ -29,6 +29,8 @@
 #include <string>
 #include <vector>
 
+#include <mata/synchronized_iterator.hh>
```
Comment (martinhruska) [inline include/mata/inter-aut.hh:32]: Why is this included here?

File: include/mata/nfa.hh [hunk 5]
```diff
@@ -148,28 +149,48 @@ Mata::Parser::ParsedSection serialize(
 	const StateToStringMap*   state_map = nullptr);
 
 
-struct TransSymbolStates {
+struct Move {
```
Comment (martinhruska) [inline include/mata/nfa.hh:152]: Generally, I don't know whether Move is it good idea to rename it to `Move`. It could be quite confusing, especially for students who know just basic definitions of FA

File: include/mata/synchronized_iterator.hh [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+            };
+
+            virtual bool advance() = 0;
+            virtual const std::vector<Iterator> get_current() = 0;
+        };
+
+
+
+        template<typename Key>
+        class SynchronizedUniverzalIterator: public SynchronizedIterator<Key> {
```
Comment (martinhruska) [inline include/mata/synchronized_iterator.hh:75]: is Univerzal an english word?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -63,7 +63,7 @@ add_library(libmata STATIC
 	nfa/nfa-segmentation.cc
 	nfa/noodlify.cc
 	rra/rrt.cc
-)
+        ../include/mata/synchronized_iterator.hh nfa/tests-synchronised_iterator.cc)
```
Comment (martinhruska) [inline src/CMakeLists.txt:66]: why is header file here?

File: src/tests-main.cc [hunk 8]
```diff
@@ -3,6 +3,9 @@
 #define CATCH_CONFIG_RUNNER
 #include "../3rdparty/catch.hpp"
 #include <mata/util.hh>
+#include <mata/util.hh>
```
Comment (martinhruska) [inline src/tests-main.cc:6]: why are these includes here?

**Review by martinhruska — COMMENTED** [review]
Beside the comments bellow I have the following issues with PR:

- You rename all occurences of the word `vector` to `_vector`. Was it intentional?
- Please remove garbage like .idea folder and especially cmake-build-debug folder.

I think that code for iterator itself is fine and can be merged as well as tests for iterator.

File: include/mata/nfa.hh (hunk 5 above)
Comment (tfiedor) [reply include/mata/nfa.hh:152]: I will also like to warn about potential clashes (mental and syntactical) with `std::move`.

File: bindings/python/mata.pxd (hunk 3 above)
Comment (kilohsakul) [reply bindings/python/mata.pxd:97]: The original name was missleading, it was not a container of Transitions but of that things which now are renamed to Move. 
I am not sure that this name is the best option, vector is not a list, and in fact I am not sure whether  we need this type name at all.

File: include/mata/inter-aut.hh (hunk 4 above)
Comment (kilohsakul) [reply include/mata/inter-aut.hh:32]: Hm, if it is not needed, then it can be removed. Or do you mean something else? Hard to tell from your question.

File: include/mata/nfa.hh (hunk 5 above)
Comment (kilohsakul) [reply include/mata/nfa.hh:152]: Well, it is better than the previous name. We anyway need to find good names for things, this is just one example. Generally, concepts that are central in the library should be named with a single word which somehow fits and sticks to the memory. Using permutations of "state" "transition" "set" for everything is silly and does not help understanding things. 
Any other idea besides "Move"?

File: 3rdparty/re2/re2/regexp.cc (hunk 1 above)
Comment (kilohsakul) [reply 3rdparty/re2/re2/regexp.cc:26]: shit. Of course not, this is nuts. I am learning to use clion, it does somewhat crazy things.

File: include/mata/synchronized_iterator.hh (hunk 6 above)
Comment (kilohsakul) [reply include/mata/synchronized_iterator.hh:75]: Do you mean that z should be s or something else? 
Indeed seem that z should be s.

File: src/tests-main.cc (hunk 8 above)
Comment (kilohsakul) [reply src/tests-main.cc:6]: I guess it can be removed.

**Comment by kilohsakul** [comment]
What is the prefered protocol, you give this kind of nice comments until I fix everything or you try to help merging things, or something in between?

File: bindings/python/mata.pxd (hunk 3 above)
Comment (kilohsakul) [reply bindings/python/mata.pxd:97]: Hm since this is the automaton transtition relation, it should have a name.

File: include/mata/inter-aut.hh (hunk 4 above)
Comment (martinhruska) [reply include/mata/inter-aut.hh:32]: You added the include by your PR so I asked what were the reasons. If it was by accident, it can be removed.

File: include/mata/nfa.hh (hunk 5 above)
Comment (tfiedor) [reply include/mata/nfa.hh:152]: I get your point, and I think it is fine. However, I would not be too "creative". I personally think, it is better to have things like `SomethingMap` for maps, `SomethingList` for lists, etc. :-).

But I guess it would be good to maybe list all of the classes/structs, identify the critical once and think if there is better name or simpify some of the names.

Comment (tfiedor) [reply include/mata/nfa.hh:152]: > What is the prefered protocol, you give this kind of nice comments until I fix everything or you try to help merging things, or something in between?

We prefer Oompa Loompas to fix their work, but since you are higher-up, you can ask for help O:-). I think, someone could help you out with the git blunders. We will squish the PR at the end, to hide the blunders, so we will have cleaner history.

File: AUTOMATAFORMAT.md (hunk 2 above)
Comment (kilohsakul) [reply AUTOMATAFORMAT.md:48]: This is shit, plus the shit with git, argh, I will better make a new branch and coppy my changes there and make another pull request. I will revert the name changes there, so that we can merge it now quickly, and we will decide on how to name things somehow soon.

**Comment by martinhruska** [comment]
The standard protocol is that an author fixes problems mentioned by reviewers. I've already fixed some minor things like unwanted includes. So please do `git pull` before conflicts arise.

**Comment by martinhruska** [comment]
Mr. @tfiedor may have an idea or two how to clean mess with debug folder and accidental `vector` rename.

**Comment by tfiedor** [comment]
> What is the prefered protocol, you give this kind of nice comments until I fix everything or you try to help merging things, or something in between?

We prefer Oompa Loompas to fix their work, but since you are higher-up, you can ask for help O:-). I think, someone could help you out with the git blunders. We will squish the PR at the end, to hide the blunders, so we will have cleaner history. (i accidentaly replied in wrong thread).

Maybe, creating new branch and new PR is the cleanest and easist solution.

**Comment by kilohsakul** [comment]
merge interator_second_service instead
